proposals/p2188.md

```diff
@@ -0,0 +1,517 @@
+    case var (n: i32, m: i32) => { F(if n then &n else &m); }
+  }
+}
+```
+
+Pattern matching precedes the initialization of the storage for any `var`
+patterns. An introduced variable is only initialized if the complete pattern
+matches.
+
+```
+class X {
+  destructor { Print("Destroyed!"); }
+}
+fn F(x: X) {
+  match ((x, 1 as i32)) {
+    case (var y: X, 0) => {}
+    case (var z: X, 1) => {}
+    // Prints "Destroyed!" only once, when `z` is destroyed.
+  }
+}
+```
+
+A `var` pattern cannot be nested within another `var` pattern. The declaration
+syntax `var` _pattern_ `=` _expresson_ `;` is equivalent to `let` `var`
+_pattern_ `=` _expression_ `;`.
+
+### Tuple patterns
+
+A tuple of patterns can be used as a pattern.
+
+-   _tuple-pattern_ ::= `(` [_expression_ `,`]\* _proper-pattern_ [`,`
+    _pattern_]\* `,`? `)`
+-   _proper-pattern_ ::= _tuple-pattern_
+
+A _tuple-pattern_ containing no commas is treated as grouping parens: the
+contained _proper-pattern_ is matched directly against the scrutinee. Otherwise,
+the behavior is as follows.
+
+A tuple pattern is matched left-to-right. The scrutinee is required to be of
+tuple type.
+
+Note that a tuple pattern must contain at least one _proper-pattern_. Otherwise,
+it is a tuple-valued expression. However, a tuple pattern and a corresponding
+tuple-valued expression are matched in the same way because `==` for a tuple
+compares fields left-to-right.
+
+### Struct patterns
+
+A struct can be matched with a struct pattern.
+
+-   _proper-pattern_ ::= `{` [_field-init_ `,`]\* _proper-field-pattern_ [`,`
+    _field-pattern_]\* `}`
+-   _proper-pattern_ ::= `{` [_field-pattern_ `,`]+ `...` `}`
+-   _field-init_ ::= _designator_ `=` _expression_
+-   _proper-field-pattern_ ::= _designator_ `=` _proper-pattern_
+-   _proper-field-pattern_ ::= _binding-pattern_
+-   _field-pattern_ ::= _field-init_
+-   _field-pattern_ ::= _proper-field-pattern_
+
+A struct pattern resembles a struct literal, with at least one field initialized
+with a proper pattern:
+
+```
+match ({.a = 1, .b = 2}) {
+  // Struct literal as an expression pattern.
+  case {.b = 2, .a = 1} => {}
+  // Struct pattern.
+  case {.b = n: i32, .a = m: i32} => {}
+}
+```
+
+The scrutinee is required to be of struct type, and to have the same set of
+field names as the pattern. The pattern is matched left-to-right, meaning that
+matching is performed in the field order specified in the pattern, not in the
+field order of the scrutinee. This is consistent with the behavior of matching
+against a tuple-valued expression, where the left operand determines the order
+in which `==` comparisons are performed.
+
+In the case where a field will be bound to an identifier with the same name, a
+shorthand syntax is available: `a: T` is synonymous with `.a = a: T`.
+
+```
+match ({.a = 1, .b = 2}) {
+  case {a: i32, b: i32} => { return a + b; }
+}
+```
+
+If some fields should be ignored when matching, a trailing `, ...` can be added
```

Comment:
I worry that this could be mistaken for a variadic syntax, or perhaps even be ambiguous with one.

As you may recall, the current (very preliminary) plan is to express pack expansion using special operator tokens that start with `...`, such as `...,` and `...{`. I don't anticipate wanting a `...}` operator, so there's no outright ambiguity, but it could easily cause confusion for human readers. Furthermore, we were previously considering using a _trailing_ `...` for those operators, in which case `,...` would be ambiguous between this syntax and the comma pack expansion. I currently think leading `...` is a better choice, but switching back does seem within the realm of possibility.

How about something like `._` as an alternative syntax? It's not perfect, because it looks like it should match exactly one field, but it's the best I've come up with so far.

OTOH I'm not very satisfied with the variadic syntax as a whole, so I'm OK with keeping this struct pattern syntax for now, in the hope that we can come up with a better variadic syntax. I just want to make sure this issue is considered.

Reply:
Could we potentially make this match the variadic syntax better, by using something like `..., _`?  That way it looks like a pack expansion of `_`, which seems like a pretty good analogy for what it's doing.

Reply:
Potentially. That seems like something to consider as part of the variadics proposal... though that syntax seems a bit cumbersome to me, and it might suggest that you could use just `, _` to ignore exactly one field, which I don't think is an interesting facility to provide.